### PR TITLE
Combine mappers

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

File: sumpy/codegen.py [hunk 1]
```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
...
     map_common_subexpression_uncached = IdentityMapper.map_common_subexpression
 
 
+def combine_mappers(*mappers):
+    """Returns a mapper that combines the work of several mappers.
+    For this to work, the mappers need to be instances of
+    :class:`sumpy.codegen.CallExternalRecMapper` and when calling
+    parent class methods, the mappers need to use the argument
+    *rec_object* passed to the *map_* method. All the mappers
+    need to be commutative and for any mapper applying the
```
Comment (inducer) [inline sumpy/codegen.py:654]:
```suggestion
    need to commute and for any mapper applying the
```

File: sumpy/codegen.py [hunk 2]
```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
...
+            if new_expr is not expr:
+                return self.rec(new_expr)
+        return expr
+
+    from functools import partial
+    import types
+    combine_mapper = CombineMapper(all_methods)
+    for method_name in all_methods.keys():
+        setattr(combine_mapper, method_name,
+                types.MethodType(partial(_map, method_name), combine_mapper))
```
Comment (inducer) [inline sumpy/codegen.py:701]: Interesting, TIL! (Could've done that in https://github.com/inducer/pytools/pull/76/files, cc @alexfikl)

File: sumpy/codegen.py [hunk 3]
```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
...
+            return getattr(IdentityMapper, method_name)(self, expr)
+        for mapper, method in self.all_methods[method_name]:
+            new_expr = method(mapper, expr, self)
+            if new_expr is not expr:
+                return self.rec(new_expr)
+        return expr
+
+    from functools import partial
+    import types
+    combine_mapper = CombineMapper(all_methods)
```
Comment (inducer) [inline sumpy/codegen.py:698]:
```suggestion
    combined_mapper = CombinedMapper(all_methods)
```

File: sumpy/codegen.py [hunk 4]
```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
...
+            self.all_methods = all_methods
+        map_common_subexpression_uncached = IdentityMapper.map_common_subexpression
+
+    def _map(method_name, self, expr, rec_object=None, *args):
+        if method_name not in self.all_methods:
+            return getattr(IdentityMapper, method_name)(self, expr)
+        for mapper, method in self.all_methods[method_name]:
+            new_expr = method(mapper, expr, self)
+            if new_expr is not expr:
+                return self.rec(new_expr)
```
Comment (inducer) [inline sumpy/codegen.py:693]:
```suggestion
                # Re-traverse the whole thing from the get-go.
                return self.rec(new_expr)
```

Comment (inducer) [inline sumpy/codegen.py:693]: Did you find this to be necessary?

File: sumpy/codegen.py [hunk 5]
```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
...
+    mapper on an expression that has already been mapped should
+    result in an identical object.
+    """
+    from collections import defaultdict
+    all_methods = defaultdict(list)
+    base_classes = [CSECachingMapperMixin, IdentityMapper]
+    for mapper in mappers:
+        assert isinstance(mapper, CallExternalRecMapper)
+        for method_name in dir(type(mapper)):
+            if not method_name.startswith("map"):
```
Comment (inducer) [inline sumpy/codegen.py:664]:
```suggestion
            if not method_name.startswith("map_"):
```

File: sumpy/codegen.py [hunk 6]
```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
...
         if expr.name == "pi":
             return prim.Variable("M_PI")
         else:
-            return IdentityMapper.map_variable(self, expr)
+            return expr
 
     map_common_subexpression_uncached = IdentityMapper.map_common_subexpression
 
 
+def combine_mappers(*mappers):
```
Comment (inducer) [inline sumpy/codegen.py:648]: It'd be nice if all the mapper-combining stuff were gathered together in one `{{{` section.

Comment (inducer) [inline sumpy/codegen.py:648]: Do you like this enough to include it in pymbolic?

File: sumpy/codegen.py [hunk 7]
```diff
@@ -518,23 +623,85 @@ def map_sum(self, expr):
...
+    """Returns a mapper that combines the work of several mappers.
+    For this to work, the mappers need to be instances of
+    :class:`sumpy.codegen.CallExternalRecMapper` and when calling
+    parent class methods, the mappers need to use the argument
+    *rec_object* passed to the *map_* method. All the mappers
```
Comment (inducer) [inline sumpy/codegen.py:653]: Maybe rename `rec_self`?

Comment (inducer) [inline sumpy/codegen.py:653]: This dispatch logic is still a bit opaque to me. Could you expand the docstring to explain how it's used?

**Review by inducer — COMMENTED** [review]
Thanks! A few thoughts below.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

**Comment by inducer** [comment]
What's the state of this? I'm looking at it now, since I need to modify this code in order to adapt to kernel callables, and I would prefer not to cause a messy conflict.

File: sumpy/codegen.py [hunk 8]
```diff
@@ -648,15 +648,13 @@ def map_variable(self, expr, *args):
...
-    all the mappers given. The mappers given need to commute and
-    for any mapper applying the mapper on an expression that has
-    already been mapped should result in an identical object.
+    """Returns a mapper that combines the work of several other mappers.  For
+    this to work, the mappers need to be instances of
+    :class:`sumpy.codegen.CallExternalRecMapper`.  When calling parent class
+    methods, the mappers need to use the (first) argument *rec_self* as the
+    instance passed to the *map_* method. *rec_self* is a (custom-generated)
+    *CombinedMapper* instance which dispatches the object to all the mappers
+    given. The mappers need to commute and be idempotent.
```
Comment (isuruf) [inline sumpy/codegen.py:657]: Thanks for updating this

**Comment by inducer** [comment]
FWIW, I don't like this code very much. I can't convince myself that it's correct, and I can't seem to convince myself that it doesn't do exponentially too much work. At the same time, I can see that it has *some* value, and it's reasonably easy to turn off. And, unfortunately, I need to dig pretty deep into this exact area to make things go with kernel callables, so I'll merge it anyhow. I hope you'll help me improve this. Let's talk about this a bit tomorrow.

**Comment by inducer** [comment]
This has considerably worsened CI times for Pytential (here by example of sumpy's downstream CI):
- [Before](https://github.com/inducer/sumpy/runs/2678220972?check_suite_focus=true) (28m)
- [After](https://github.com/inducer/sumpy/runs/2714092339?check_suite_focus=true) (2h)

I'll back out use of the combined mappers in a new PR.

cc @kaushikcfd

**Pull request merged**